lsps2/intercept_handler.go

```diff
@@ -0,0 +1,732 @@
+	id               string
+	fakeScid         lightning.ShortChannelID
+	incomingSumMsat  uint64
+	outgoingSumMsat  uint64
+	paymentSizeMsat  uint64
+	feeMsat          uint64
+	registration     *BuyRegistration
+	parts            map[string]*partState
+	isFinal          bool
+	timoutChanClosed bool
+	timeoutChan      chan struct{}
+}
+
+func (p *paymentState) closeTimeoutChan() {
+	if p.timoutChanClosed {
+		return
+	}
+
+	close(p.timeoutChan)
+	p.timoutChanClosed = true
+}
+
+type partState struct {
+	isProcessed bool
+	isFinalized bool
+	req         *shared.InterceptRequest
+	resolution  chan *shared.InterceptResult
+}
+
+type registrationReadyEvent struct {
+	paymentId    string
+	registration *BuyRegistration
+}
+
+type paymentChanOpenedEvent struct {
+	paymentId       string
+	scid            lightning.ShortChannelID
+	channelPoint    *wire.OutPoint
+	htlcMinimumMsat uint64
+}
+
+type paymentFailureEvent struct {
+	paymentId string
+	code      shared.InterceptFailureCode
+}
+
+type awaitingRegistrationEvent struct {
+	paymentId string
+	partId    string
+}
+
+func (i *Interceptor) Start(ctx context.Context) {
+	// Main event loop for stages of htlcs to be handled. Note that the event
+	// loop has to execute quickly, so any code running in the 'handle' methods
+	// must execute quickly. If there is i/o involved, or any waiting, run that
+	// code in a goroutine, and place an event onto the event loop to continue
+	// processing after the slow operation is done.
+	// The nice thing about an event loop is that it runs on a single thread.
+	// So there's no locking needed, as long as everything that needs
+	// synchronization goes through the event loop.
+	for {
+		select {
+		case part := <-i.newPart:
+			i.handleNewPart(part)
+		case ev := <-i.registrationReady:
+			i.handleRegistrationReady(ev)
+		case paymentId := <-i.notRegistered:
+			i.handleNotRegistered(paymentId)
+		case ev := <-i.partAwaitingRegistration:
+			i.handlePartAwaitingRegistration(ev)
+		case paymentId := <-i.paymentReady:
+			i.handlePaymentReady(paymentId)
+		case paymentId := <-i.paymentTimeout:
+			i.handlePaymentTimeout(paymentId)
+		case ev := <-i.paymentFailure:
+			i.handlePaymentFailure(ev.paymentId, ev.code)
+		case ev := <-i.paymentChanOpened:
+			i.handlePaymentChanOpened(ev)
+		}
+	}
+}
+
+func (i *Interceptor) handleNewPart(part *partState) {
+	payment, paymentExisted := i.inflightPayments[part.req.PaymentId()]
+	if !paymentExisted {
+		payment = &paymentState{
+			id:          part.req.PaymentId(),
+			fakeScid:    part.req.Scid,
+			parts:       make(map[string]*partState),
+			timeoutChan: make(chan struct{}),
+		}
+		i.inflightPayments[part.req.PaymentId()] = payment
+
+		go func() {
+			select {
+			case <-time.After(i.config.MppTimeout):
+				// Handle timeout inside the event loop, to make sure there are
+				// no race conditions, since this timeout watcher is running in
+				// a goroutine.
+				i.paymentTimeout <- part.req.PaymentId()
```

Comment:
Perhaps use directly the paymentFailure channel here in favor of omitting the paymentTimeout channel?

Reply:
Done